From 2eba28c22b1edad3e9d30e1c1da1ea1694b1a233 Mon Sep 17 00:00:00 2001
From: Miroslav Grepl <mgrepl@redhat.com>
Date: Mon, 30 Jun 2014 14:58:10 +0200
Subject: [PATCH 2/2] avc_has_perm will now return yes if the machine is in
 permissive mode

---
 libselinux/src/avc.c | 8 ++++++--
 1 file changed, 6 insertions(+), 2 deletions(-)

diff --git a/libselinux/src/avc.c b/libselinux/src/avc.c
index e9e3529..ea2d208 100644
--- a/libselinux/src/avc.c
+++ b/libselinux/src/avc.c
@@ -772,11 +772,15 @@ int avc_has_perm_noaudit(security_id_t ssid,
 			rc = security_compute_av_flags_raw(ssid->ctx, tsid->ctx,
 							   tclass, requested,
 							   &entry.avd);
-			if (rc)
+			if (rc && errno == EINVAL && !avc_enforcing) {
+                rc = errno = 0;
 				goto out;
-			rc = avc_insert(ssid, tsid, tclass, &entry, aeref);
+            }
 			if (rc)
 				goto out;
+			rc = avc_insert(ssid, tsid, tclass, &entry, aeref);
+            if (rc)
+                goto out;
 		}
 		ae = aeref->ae;
 	}
-- 
2.0.0

