--- libselinux-2.0.94/src/label.c~	2013-04-25 08:43:13.591724318 +0200
+++ libselinux-2.0.94/src/label.c	2013-05-07 12:20:17.270836449 +0200
@@ -52,13 +52,21 @@
 static char *selabel_sub(const char *src) 
 {
 	char *dst = NULL;
+    int len;
 	SELABELSUB *ptr = selabelsublist;
+
 	while (ptr) {
 		if (strncmp(src, ptr->src, ptr->slen) == 0 ) {
 			if (src[ptr->slen] == '/' || 
 			    src[ptr->slen] == 0) {
-				asprintf(&dst, "%s%s", ptr->dst, &src[ptr->slen]);
-				return dst;
+                if ((src[ptr->slen] == '/') && 
+                    (strcmp(ptr->dst, "/") == 0))
+                    len = ptr->slen + 1;
+                else
+                    len = ptr->slen;
+				if (asprintf(&dst, "%s%s", ptr->dst, &src[len]) < 0)
+                    return NULL;
+			    return dst;
 			}
 		}
 		ptr = ptr->next;
