diff -up libselinux-2.0.94/src/selinuxswig_python.i.realpath libselinux-2.0.94/src/selinuxswig_python.i
--- libselinux-2.0.94/src/selinuxswig_python.i.realpath	2010-03-24 15:52:20.000000000 -0400
+++ libselinux-2.0.94/src/selinuxswig_python.i	2011-05-25 13:18:22.090140862 -0400
@@ -12,8 +12,15 @@ import shutil, os, stat
 
 def restorecon(path, recursive=False):
     """ Restore SELinux context on a given path """
-    mode = os.lstat(path)[stat.ST_MODE]
-    status, context = matchpathcon(path, mode)
+
+    try:
+        mode = os.lstat(path)[stat.ST_MODE]
+        status, context = matchpathcon(path, mode)
+    except OSError:
+        path = os.path.realpath(os.path.expanduser(path))
+        mode = os.lstat(path)[stat.ST_MODE]
+        status, context = matchpathcon(path, mode)
+
     if status == 0:
         lsetfilecon(path, context)
         if recursive:
@@ -21,6 +28,14 @@ def restorecon(path, recursive=False):
                              map(restorecon, [os.path.join(dirname, fname)
                                               for fname in fnames]), None)
 
+def chcon(path, context, recursive=False):
+    """ Set the SELinux context on a given path """
+    lsetfilecon(path, context)
+    if recursive:
+        for root, dirs, files in os.walk(path):
+            for name in files + dirs:
+               lsetfilecon(os.path.join(root,name), context)
+
 def copytree(src, dest):
     """ An SELinux-friendly shutil.copytree method """
     shutil.copytree(src, dest)
