diff --git a/src/label_file.c b/src/label_file.c
index 509b033..8ab9902 100644
--- a/src/label_file.c
+++ b/src/label_file.c
@@ -395,6 +395,8 @@ static int process_line(struct selabel_handle *rec,
 	return 0;
 }
 
+static void closef(struct selabel_handle *rec);
+
 static int init(struct selabel_handle *rec, struct selinux_opt *opts,
 		unsigned n)
 {
@@ -536,12 +538,15 @@ static int init(struct selabel_handle *rec, struct selinux_opt *opts,
 	status = 0;
 finish:
 	fclose(fp);
-	if (data->spec_arr != spec_copy)
-		free(data->spec_arr);
 	if (homedirfp)
 		fclose(homedirfp);
 	if (localfp)
 		fclose(localfp);
+	if (status) {
+		closef(rec);
+		rec->data = NULL;
+	}
+
 	return status;
 }
 
